pkg/frontend/transport/handler.go

```diff
@@ -252,6 +265,97 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getQueryDetail(url string, values url.Values) QueryDetail {
+	urlValues := parseURLValues(values)
+	t := InstantQuery
+	if strings.HasSuffix(url, "query_range") {
+		t = RangeQuery
+	}
+	return QueryDetail{
+		QueryString: urlValues.query,
+		StepInt:     urlValues.stepInt,
+		StartInt:    urlValues.startInt,
+		EndInt:      urlValues.endInt,
+		TsInt:       urlValues.tsInt,
+		QueryType:   t,
+	}
+}
+
+type urlValues struct {
+	query                            string
+	stepInt, startInt, endInt, tsInt int64
+}
+
+func parseURLValues(values url.Values) urlValues {
+	query := values.Get("query")
+	step := values.Get("step")
+	start := values.Get("start")
+	end := values.Get("end")
+	ts := values.Get("time")
+	var (
+		tsInt, startInt, endInt, stepInt int64
+		err                              error
+	)
+	if len(ts) > 0 {
+		tsInt, err = util.ParseTime(ts)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse time", "err", err)
+		}
+	}
+	if len(start) > 0 {
+		startInt, err = util.ParseTime(start)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse start", "err", err)
+		}
+	}
+
+	if len(end) > 0 {
+		endInt, err = util.ParseTime(end)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse end", "err", err)
+		}
+	}
+
+	if len(step) > 0 {
+		stepInt, err = util.ParseDurationMs(step)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse step", "err", err)
+		}
+	}
+	return urlValues{
+		query:    query,
+		tsInt:    tsInt,
+		startInt: startInt,
+		endInt:   endInt,
+		stepInt:  stepInt,
+	}
+}
+
+type queryOrigin struct{}
+
+type QueryDetail struct {
+	QueryString      string
+	SamplesProcessed int64
+	QueryType        QueryType
+	StartInt         int64
+	EndInt           int64
+	StepInt          int64
+	TsInt            int64
+}
+
+// NewOriginContext returns a new context with data about the origin attached.
+func NewOriginContext(ctx context.Context, query QueryDetail) context.Context {
+	return context.WithValue(ctx, queryOrigin{}, query)
+}
+
+// FromOriginContext returns the QueryDetail origin data from the context.
+func FromOriginContext(ctx context.Context) QueryDetail {
+	if queryInfo, ok := ctx.Value(queryOrigin{}).(QueryDetail); ok {
+		return queryInfo
+	}
+	return QueryDetail{}
+}
```

Comment:
I don't think we need `NewOriginContext` and `FromOriginContext` in Cortex. They are unused here, right